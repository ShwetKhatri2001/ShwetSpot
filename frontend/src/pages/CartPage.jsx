import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import {
  Button,
  Card,
  Col,
  Form,
  Image,
  ListGroup,
  Row,
} from 'react-bootstrap';
import Message from '../components/Message';
import { addToCart, removeFromCart } from '../redux/actions/cartActions';

const CartPage = ({ match, location, history }) => {
  const productId = match.params.id;

  const qty = location.search ? Number(location.search.split('=')[1]) : 1;

  const dispatch = useDispatch();

  const cart = useSelector(state => state.cart);
  const { cartItems } = cart;

  useEffect(() => {
    if (productId) {
      dispatch(addToCart(productId, qty));
    }
  }, [dispatch, productId, qty]);

  const removeFromCartHandler = id => {
    dispatch(removeFromCart(id));
  };

  const checkoutHandler = () => {
    history.push('/login?redirect=shipping');
  };

  return (
    <Row>
      <Col md={8}>
        <h1>Shopping Cart</h1>
        {cartItems.length === 0 ? (
          <Message>
            Your cart is empty <Link to="/">Go Back</Link>
          </Message>
        ) : (
          <ListGroup>
            {cartItems.map(item => (
              <ListGroup.Item key={item.product}>
                <Row>
                  <Col md={2}>
                    <Image
                      src={item.image}
                      alt={item.name}
                      rounded
                      fluid
                      id="cart-image"
                    />
                  </Col>
                  <Col md={3}>
                    <Link to={`/product/${item.product}`}>{item.name}</Link>
                  </Col>
                  <Col md={2}>{item.price} credits</Col>
                  <Col md={2}>
                    <Form.Control
                      as="select"
                      className="custom-select custom-select-m"
                      value={item.qty}
                      onChange={e =>
                        dispatch(
                          addToCart(item.product, Number(e.target.value))
                        )
                      }
                    >
                      {[...Array(item.countInStock).keys()].map(x => (
                        <option key={x + 1} value={x + 1}>
                          {x + 1}
                        </option>
                      ))}
                    </Form.Control>
                  </Col>
                  <Col md={3} className="text-right">
                    <Button
                      type="button"
                      variant="danger"
                      onClick={() => removeFromCartHandler(item.product)}
                    >
                      <i className="fas fa-trash" />
                    </Button>
                  </Col>
                </Row>
              </ListGroup.Item>
            ))}
          </ListGroup>
        )}
      </Col>
      <Col md={4}>
        <Card>
          <ListGroup variant="flush">
            <ListGroup.Item>
              <h2>
                Subtotal ({cartItems.reduce((acc, item) => acc + item.qty, 0)})
                items
              </h2>
              {cartItems
                .reduce((acc, item) => acc + item.qty * item.price, 0)
                .toFixed(2)}{' '}
              credits
            </ListGroup.Item>
            <ListGroup.Item>
              <Button
                type="button"
                className="btn-block"
                variant="success"
                disabled={cartItems.length === 0}
                onClick={checkoutHandler}
              >
                Proceed To Checkout
              </Button>
            </ListGroup.Item>
          </ListGroup>
        </Card>
      </Col>
    </Row>
  );
};

export default CartPage;
