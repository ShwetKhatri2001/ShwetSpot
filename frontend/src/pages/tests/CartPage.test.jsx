import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import CartPage from '../CartPage';
import store from '../../redux/store';

describe('CartPage Component', () => {
  const MockCartPage = (
    <MemoryRouter>
      <Provider store={store}>
        <CartPage match={{ params: { id: 1 } }} location={{ search: '1' }} />
      </Provider>
    </MemoryRouter>
  );

  it('should match the snapshot', () => {
    const tree = renderer.create(MockCartPage).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
