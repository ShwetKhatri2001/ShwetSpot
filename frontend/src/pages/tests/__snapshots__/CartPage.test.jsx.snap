// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPage Component should match the snapshot 1`] = `
<div
  className="row"
>
  <div
    className="col-md-8"
  >
    <h1>
      Shopping Cart
    </h1>
    <div
      className="fade alert alert-info show"
      role="alert"
    >
      Your cart is empty 
      <a
        href="/"
        onClick={[Function]}
      >
        Go Back
      </a>
    </div>
  </div>
  <div
    className="col-md-4"
  >
    <div
      className="card"
    >
      <div
        className="list-group list-group-flush"
        onKeyDown={[Function]}
      >
        <div
          className="list-group-item"
          data-rb-event-key={null}
          disabled={false}
          onClick={[Function]}
        >
          <h2>
            Subtotal (
            0
            ) items
          </h2>
          0.00
           
          credits
        </div>
        <div
          className="list-group-item"
          data-rb-event-key={null}
          disabled={false}
          onClick={[Function]}
        >
          <button
            className="btn-block btn btn-success"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Proceed To Checkout
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
